[MRG+1] DEP Change default of error_score in cross-validation

scouvreur opened the pull request:
#### Reference Issues/PRs

This was raised by @glemaitre in issue #13797 as issues to resolve before the v0.22 release.

#### What does this implement/fix? Explain your changes.

- [x] Changed default of `error_score` in model_selection from 'raise-deprecating' to np.nan in :
    - [x] model_selection._search
    - [x] model_selection._validation
- [x] Removed test for deprecation message on v0.22 in test_validation

scouvreur committed:
Issue #13797: Change error_score default

Changed default of `error_score` in model_selection from
'raise-deprecating' to np.nan in model_selection._search
and model_selection._validation. Removed test for deprecation
message on v0.22 in test_validation.

On branch change_default_error-score_validation_tools
Changes to be committed:
	modified:   _search.py
	modified:   _validation.py
	modified:   tests/test_validation.py

scouvreur committed:
[WIP] Issue #13797: Change error_score default

Merge remote-tracking branch 'upstream/master'
into change_default_error-score_validation_tools.

glemaitre commented:
Could you remove the following explicit affectations `error_score=np.nan` or something similar:

* in `test_multioutput.py` l. 192
* in `test_search.py` l. 1465

glemaitre reviewed (approved):
Otherwise LGTM

scouvreur committed:
[MRG] Issue #13797: Change error_score default

Removed instances where `error_score` was set to 0,
float('nan'), np.nan or 0.0 as the new default is
np.nan in v0.22 in test_validation.

On branch change_default_error-score_validation_tools
Your branch is up to date with 'origin/change_default_error-score_validation_tools'.

Changes to be committed:
	modified:   sklearn/model_selection/tests/test_search.py
	modified:   sklearn/tests/test_multioutput.py

glemaitre commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1446,7 +1446,7 @@ def test_grid_search_failing_classifier():
     # error in this test.
     gs = GridSearchCV(clf, [{'parameter': [0, 1, 2]}], scoring='accuracy',
-                      refit=False, error_score=0.0)
+                      refit=False)
```
revert this change (we need to remove only when error_score=np.nan)

glemaitre commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1242,8 +1242,8 @@ def test_search_cv_timing():
 
-    gs = GridSearchCV(svc, {'C': [0, 1]}, cv=2, error_score=0)
-    rs = RandomizedSearchCV(svc, {'C': [0, 1]}, cv=2, error_score=0, n_iter=2)
+    gs = GridSearchCV(svc, {'C': [0, 1]}, cv=2)
```
revert this change (we need to remove only when error_score=np.nan)

glemaitre commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1242,8 +1242,8 @@ def test_search_cv_timing():
-    gs = GridSearchCV(svc, {'C': [0, 1]}, cv=2, error_score=0)
-    rs = RandomizedSearchCV(svc, {'C': [0, 1]}, cv=2, error_score=0, n_iter=2)
+    gs = GridSearchCV(svc, {'C': [0, 1]}, cv=2)
+    rs = RandomizedSearchCV(svc, {'C': [0, 1]}, cv=2, n_iter=2)
```
revert this change (we need to remove only when error_score=np.nan)

glemaitre reviewed (dismissed):
This is the cause of the failure

scouvreur committed:
Reverted changes on test_search

Reverted changes on test_search for when
error_score=0 or 0.0 was passed.

On branch change_default_error-score_validation_tools
Your branch is up to date with 'origin/change_default_error-score_validation_tools'.

Changes to be committed:
	modified:   test_search.py

scouvreur committed:
Change error_score default

Merge remote-tracking branch 'upstream/master'
into change_default_error-score_validation_tools.

glemaitre commented:
Still failing

scouvreur commented:
Hi @glemaitre, I was trying to understand the test failure in azure pipelines - it seems to be with the default of cv='warn' in check_cv. Would you like me to change the default there also ?

NicolasHug commented on sklearn/model_selection/_search.py:
```diff
@@ -329,8 +329,7 @@ def fit_grid_point(X, y, estimator, parameters, train, test, scorer,
         FitFailedWarning is raised. This parameter does not affect the refit
-        step, which will always raise the error. Default is 'raise' but from
-        version 0.22 it will change to np.nan.
+        step, which will always raise the error.
```
Please indicate the default value in the description: "default is np.nan"

Here and in the relevant other places

NicolasHug commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1479,7 +1479,6 @@ def get_cand_scores(i):
     assert gs.best_index_ != clf.FAILING_PARAMETER
 
 
-@pytest.mark.filterwarnings('ignore: The default value of cv')  # 0.22
```
don't remove this, this is not related. That's why CI is failing

NicolasHug commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1666,16 +1666,6 @@ def test_fit_and_score_failing():
-                       "error_score='raise' if you want an exception "
-                       "raised or error_score=np.nan to adopt the "
-                       "behavior from version 0.22.")
-    with pytest.raises(ValueError):
```
I think we still want the `with pytest.raises(ValueError):` here.

The `assert_warns_message` is the only part that should be removed:

```py
with pytest.raises(ValueError):
    _fit_and_score(*fit_and_score_args)
```

scouvreur replied:
Thanks for the feedback ! Somehow it seems that I still get a test failure from `test_fit_and_score_failing()`, could it be related due to this structure in `_validation.py` ?

```python
except Exception as e:
    # Note fit time as time until error
    fit_time = time.time() - start_time
    score_time = 0.0
    if error_score == 'raise':
        raise
    elif isinstance(error_score, numbers.Number):
        if is_multimetric:
            test_scores = dict(zip(scorer.keys(),
                               [error_score, ] * n_scorers))
            if return_train_score:
                train_scores = dict(zip(scorer.keys(),
                                    [error_score, ] * n_scorers))
        else:
            test_scores = error_score
            if return_train_score:
                train_scores = error_score
        warnings.warn("Estimator fit failed. The score on this train-test"
                      " partition for these parameters will be set to %f. "
                      "Details: \n%s" %
                      (error_score, format_exception_only(type(e), e)[0]),
                      FitFailedWarning)
    else:
        raise ValueError("error_score must be the string 'raise' or a"
                         " numeric value. (Hint: if using 'raise', please"
                         " make sure that it has been spelled correctly.)")
```

NicolasHug replied:
My bad, I did not realize that the previous default would also raise ValueError.

You can remove it as you previously did.

(in the future please avoid mentioning with @username in the git commit messages, because it will send notifications everytime you commit which is not necessarily wanted ;) )

scouvreur replied:
Got it - apologies for that ! Hopefully this will be the last commit in this PR and then the notifications will stop, noted for future reference.

scouvreur committed:
Added default documentation to docstrings

Added default is np.nan to fit_grid_point()
function, and GridSearchCV(), RandomizedSearchCV()
classes.

On branch change_default_error-score_validation_tools
Your branch is up to date with 'origin/change_default_error-score_validation_tools'.

Changes to be committed:
	modified:   model_selection/_search.py

scouvreur committed:
Updated failing tests

Updated failing tests according to review from
@NicolasHug.

On branch change_default_error-score_validation_tools
Your branch is ahead of 'origin/change_default_error-score_validation_tools' by 1 commit.
   (use "git push" to publish your local commits)

Changes to be committed:
	modified:   test_search.py
	modified:   test_validation.py

scouvreur committed:
Removed value error test

Removed value error test in test_fit_and_score_failing().

On branch change_default_error-score_validation_tools
Your branch is up to date with 'origin/change_default_error-score_validation_tools'.

Changes to be committed:
	modified:   test_validation.py

NicolasHug commented:
Please also update docstrings (no need to touch test_pprint though):

```
~/dev/sklearn(branch:pr-13840*) » git grep raise-deprecating 
sklearn/model_selection/_validation.py:    error_score : 'raise' | 'raise-deprecating' or numeric
sklearn/model_selection/_validation.py:        If set to 'raise-deprecating', a FutureWarning is printed before the
sklearn/model_selection/_validation.py:    error_score : 'raise' | 'raise-deprecating' or numeric
sklearn/model_selection/_validation.py:        If set to 'raise-deprecating', a FutureWarning is printed before the
sklearn/model_selection/_validation.py:    error_score : 'raise' | 'raise-deprecating' or numeric
sklearn/model_selection/_validation.py:        If set to 'raise-deprecating', a FutureWarning is printed before the
sklearn/model_selection/_validation.py:    error_score : 'raise' | 'raise-deprecating' or numeric
sklearn/model_selection/_validation.py:        If set to 'raise-deprecating', a FutureWarning is printed before the
sklearn/model_selection/_validation.py:    error_score : 'raise' | 'raise-deprecating' or numeric
sklearn/model_selection/_validation.py:        If set to 'raise-deprecating', a FutureWarning is printed before the
sklearn/utils/tests/test_pprint.py:                 pre_dispatch='2*n_jobs', error_score='raise-deprecating',
sklearn/utils/tests/test_pprint.py:GridSearchCV(cv=5, error_score='raise-deprecating',
sklearn/utils/tests/test_pprint.py:GridSearchCV(cv=3, error_score='raise-deprecating',
sklearn/utils/tests/test_pprint.py:GridSearchCV(cv='warn', error_score='raise-deprecating',
sklearn/utils/tests/test_pprint.py:GridSearchCV(cv='warn', error_score='raise-deprecating',
```

scouvreur committed:
Removed 'raise-deprecating' from docstrings

On branch change_default_error-score_validation_tools
Your branch is up to date with 'origin/change_default_error-score_validation_tools'.

Changes to be committed:
	modified:   ../_validation.py

NicolasHug reviewed (approved):
LGTM, thanks scouvreur

@glemaitre I think your comments were addressed

jnothman reviewed (approved):
Thanks @scouvreur

scouvreur commented:
Thanks guys - learned alot from working on this !

glemaitre commented:
Thanks @scouvreur. One less!!! :)
